### DISCUSSION TITLE
How can I insert pagebreak for headings living at the bottom of the page (using R Quarto + typst , output: PDF)?

### DISCUSSION THREAD

**Discussion by haokaitseng** [quarto-dev/quarto-cli#11591] [Q&A]
### Description

Dear Quarto community,

I am using Quarto in R to generate PDFs, based on the idea of the newsletter typst [template](https://github.com/quarto-ext/typst-templates/tree/main/dept-news) developed by [mskyttner](https://github.com/quarto-ext/typst-templates/commits?author=mskyttner).

The pain point is that I don't want the heading living at the bottom of the page, I want to enforce a page break before the heading, so that the heading can go to the next new page along with the contents (see below screenshot). 

<img width="697" alt="image" src="https://github.com/user-attachments/assets/115689ff-b407-4e6f-ae0f-8dcf4bf3366f">

I also visited relevant discussions, such as [this](https://github.com/quarto-dev/quarto-cli/discussions/11164) and [this](https://github.com/quarto-dev/quarto-cli/discussions/7060), and [this](https://typst.app/docs/reference/layout/pagebreak/). They talked about pagebreak() between different components, but what I want is page breaks within content body itself. And pagebreak(), \pagebreak, and {{< pagebreak >}} all didn't work of course. 

I am thinking, take below as example, 
- strategy 1: if I treat the content as two bodies, so that I can force a pagebreak() between the first body and the second body in my typst template.
- strategy 2: Since the length of each parametrized tables are changing, is there any condition that I can use to detect, only when the heading sit at the bottom of the page then I enforce it move it into the next page?

My current QMD structure looks like below:

````
---
format: 
  info-sheet-typst: default   
 title: "XXX"
edition: "XXX"
--- 
# heading 1
table1 
# heading 2
table2 
**(what I want is a page break here)**
# heading 3
table3 
# heading 4
table4 
````
Many thanks for thoughts and suggestions! Also thanks again for support the great typst product running in Quarto! 
Best wishes,
Kai

**Comment by mcanouil** [comment]
Could you properly format your post using code blocks for code and terminal outputs? Thanks.
If your code contains code blocks, you need to enclose it using more backticks, *i.e.*, usually four ` ```` `.
See <https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us>.


What do you mean `{{< pagebreak >}}` does not work? It should. https://quarto.org/docs/authoring/markdown-basics.html#page-breaks

**Reply by ironupiwada** [reply]
It won’t. @haokaitseng clearly stated that the issue concerns Typst – and indeed, compiling to Typst PDF with Quarto, with the `{{< pagebreak >}}` shortcode in the content, results in the error: "pagebreaks are not allowed inside of containers." I have a feeling that the shortcode [is not natively supported](https://quarto.org/docs/authoring/markdown-basics.html#page-breaks) in Typst PDF output—hopefully, at least until the Typst 0.12 release support.

**Reply by mcanouil** [reply]
The shortcode works just fine, as requested previously, we need a reproducible example (*i.e.*, commenting "It won’t." does not help us help you or anyone).

<img width="1624" alt="image" src="https://github.com/user-attachments/assets/a8db3443-28bc-4764-a954-e1370f5cec6a" />

````qmd
---
title: "Quarto Playground"
format: typst
---

This is a playground for Quarto.

{{< lipsum 1 >}}

{{< pagebreak >}}

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}
````

**Reply by mcanouil** [reply]
Note that here the original user is using what seems to be a custom format.

**Reply by ironupiwada** [reply]
The answer "It should" doesn’t help either. Consider this code:
```
---
title: "Quarto Playground"
format: 
    typst:
      toc: true
      columns: 2
---

This is a playground for Quarto.

{{< lipsum 1 >}}

{{< pagebreak >}}
```

With Quarto 1.7.7, the log is following:

```
pandoc 
  to: typst
  output-file: Untitled-1.typ
  standalone: true
  shift-heading-level-by: -1
  default-image-extension: svg
  wrap: none
  citeproc: false
  
metadata
  columns: 2
  title: Quarto Playground
  
[typst]: Compiling Untitled-1.typ to Untitled-1.pdf...error: pagebreaks are not allowed inside of containers
    ┌─ Untitled-1.typ:304:1
    │
304 │ #pagebreak()
    │  ^^^^^^^^^^^
```

After deleting the ```toc: true``` option in the YAML preamble, it seems that the problem is with `columns`.

**Reply by mcanouil** [reply]
> The answer "It should" doesn’t help either. Consider this code:

"it should" was associated with a link to the documentation showing how to use it. As you can see, I did not add much into it to show it works. (note that the screenshot posted does not use multiple columns, so a different use cases than yours)

This being said, thanks for the example with the multiple columns layout. This is helpful.

**Reply by mcanouil** [reply]
I've opened an enhancement issue but I believe the "issue" to be upstream:

- https://github.com/quarto-dev/quarto-cli/issues/11819

**Reply by mcanouil** [reply]
Note that in this case you should use `` `#colbreak()`{=typst} ``.

To be clear, this is not a Quarto or Pandoc bug, this is Typst behaviour by design (`colbreak`/`pagebreak`).

````qmd
---
title: "Quarto Playground"
format: 
  typst:
    columns: 2
---

This is a playground for Quarto.

{{< lipsum 1 >}}

`#colbreak()`{=typst}

{{< lipsum 1 >}}

`#colbreak()`{=typst}

{{< lipsum 5 >}}
````

**Reply by haokaitseng** [reply]
Dear mcanouil and ironupiwada,
Thank you so so so much for investageting and supporting this issue.
Yes, the template that I used, [quarto-ext/typst-templates](https://github.com/quarto-ext/typst-templates/tree/main/dept-news), has a left and a right columns. 

In my screen shot, the blue color strip is the left column side bar. The table is the main text column.

Let me try to remove the bar slide bar column and see if that works.
Apologies for me not able to provide a reproducible example, and thanks for your patience.
Thanks again for this hint. 
 
 Best regards,
Kai

**Reply by mcanouil** [reply]
This template uses columns in it, so you can't use a "page break" instead you need a column break as illustrated before.
This is a Typst requirement/constraint.


Quarto can only abstract "basic features" across formats, beyond that you have no other choice than reading/learning the underlying technology's documentation:
- `format: pdf` -> LaTeX
- `format: typst` -> Typst
- `format: docx` -> OpenXML
- `format: html` -> HTML/CSS
- *etc.*
